File: leetcode/872.cpp
#include <vector>
#include <algorithm>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
private:
  void dfs(TreeNode *root, vector<int> &result) {
    if (root == nullptr) {
      return;
    }
    if (root->left == nullptr && root->right == nullptr) {
      result.push_back(root->val);
      return;
    }
    dfs(root->left, result);
    dfs(root->right, result);
  }

public:
  bool leafSimilar(TreeNode *root1, TreeNode *root2) {
    vector<int> leaf1;
    vector<int> leaf2;
    dfs(root1, leaf1);
    dfs(root2, leaf2);
    // return equal(leaf1.begin(), leaf1.end(), leaf2.begin(), leaf2.end());
    return leaf1 == leaf2;
  }
};

File: leetcode/872.py
from typing import List


# Definition for a binary tree node.
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:

        def dfs(root: TreeNode, ans: List[TreeNode]):
            if not root:
                return

            if (not root.left) and (not root.right):
                ans.append(root.val)
                return

            dfs(root.left, ans)
            dfs(root.right, ans)

        ans1, ans2 = [], []
        dfs(root1, ans1)
        dfs(root2, ans2)
        return ans1 == ans2
